File: geocoding/src/main/java/se/arbitur/geocoding/Geocoder.java
package se.arbitur.geocoding;

import android.os.Handler;
import android.os.Looper;

import com.google.gson.Gson;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.logging.HttpLoggingInterceptor;
import se.arbitur.geocoding.Constants.ResponseStatuses;



public abstract class Geocoder {
	public static HttpLoggingInterceptor.Level loggingLevel = HttpLoggingInterceptor.Level.BODY;

	protected static HttpUrl.Builder getDefaultUrlBuilder() {
		return new HttpUrl.Builder()
				.scheme("https")
				.host("maps.googleapis.com")
				.addPathSegment("maps")
				.addPathSegment("api")
				.addPathSegment("geocode")
				.addPathSegment("json");
	}

	protected String key;
	protected String language;



	protected Geocoder(String key) {
		this.key = key;
	}



	protected abstract HttpUrl.Builder getUrlBuilder();



	public void fetch(final Callback callback) {
		OkHttpClient httpClient = new OkHttpClient.Builder()
				.addInterceptor(new HttpLoggingInterceptor().setLevel(loggingLevel))
				.build();


		HttpUrl.Builder url = getUrlBuilder();

		if (language != null) 	url.addQueryParameter("language", language);
		if (key != null) 		url.addQueryParameter("key", key);

		Request request = new Request.Builder()
				.url(url.build())
				.build();

		httpClient.newCall(request).enqueue(new okhttp3.Callback() {
			Handler handler = new Handler(Looper.getMainLooper());

			@Override
			public void onFailure(Call call, final IOException e) {
				handler.post(new Runnable() {
					@Override
					public void run() {
						callback.onFailed(null, e);
					}
				});
			}

			@Override
			public void onResponse(Call call, okhttp3.Response response) throws IOException {
				String rawBody = response.body().string();
				final Response geoResponse = new Gson().fromJson(rawBody, Response.class);

				if (geoResponse.getStatus().equals(ResponseStatuses.OK)) {
					handler.post(new Runnable() {
						@Override
						public void run() {
							callback.onResponse(geoResponse);
						}
					});
				}
				else {
					handler.post(new Runnable() {
						@Override
						public void run() {
							callback.onFailed(geoResponse, null);
						}
					});
				}
			}
		});
	}
}

File: geocoding/src/main/java/se/arbitur/geocoding/Geocoder.kt
package se.arbitur.geocoding


import android.os.Handler
import android.os.Looper
import okhttp3.Call
import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.logging.HttpLoggingInterceptor
import org.json.JSONObject
import se.arbitur.geocoding.constants.ResponseStatus
import java.io.IOException
import se.arbitur.geocoding.Response as GeoResponse



abstract class Geocoder(val key: String?) {

	companion object {
		@JvmField var LOGGING_LEVEL = HttpLoggingInterceptor.Level.BODY

		@JvmStatic protected val defaultUrlBuilder: HttpUrl.Builder
			get() = HttpUrl.Builder()
							.scheme("https")
							.host("maps.googleapis.com")
							.addPathSegment("maps")
							.addPathSegment("api")
							.addPathSegment("geocode")
							.addPathSegment("json")
	}



	protected var language: String? = null

	protected abstract val urlBuilder: HttpUrl.Builder

    private val looper = Handler(Looper.getMainLooper())


	fun fetch(callback: Callback) {
		val httpClient = OkHttpClient.Builder()
				.addInterceptor(HttpLoggingInterceptor().setLevel(LOGGING_LEVEL))
				.build()

		val url = urlBuilder

		if (key != null) url.addQueryParameter("key", key)
		if (language != null) url.addQueryParameter("language", language)

		val request = Request.Builder()
				.url(url.build())
				.build()

		httpClient.newCall(request).enqueue(object : okhttp3.Callback {
			override fun onResponse(call: Call?, response: Response?) {
				try {
					val json = response?.body()?.string() ?: throw NullPointerException()
					val resp = GeoResponse(JSONObject(json))

					if (resp.status == ResponseStatus.OK)
                        looper.post { callback.onSuccess(resp) }
					else
                        looper.post { callback.onFailure(resp, null) }
				}
				catch (ex: Exception) {
					throw IOException(ex)
				}
			}

			override fun onFailure(call: Call?, e: IOException?) {
                looper.post { callback.onFailure(null, e) }
			}
		})
	}
}
